File: focusfire.lua
-- Configuration
local CONFIG = {
    minAttackers = 2,  -- Minimum number of different attackers required to highlight target
    trackerTimeWindow = 4.5,  -- How long to consider someone as being "targeted" after being shot
    cleanupInterval = 0.5,  -- How often to clean up invalid targets (seconds)
    
    -- Visualization modes
    enableChams = true,  -- Enable chams visualization
    enableBox = true,  -- Enable border box visualization
    visibleOnly = false, -- Only show effects on visible players
    
    -- Box settings
    boxColor = {r = 255, g = 0, b = 0, a = 190},  -- RGBA color for the box
    boxThickness = 2,  -- Thickness of the box border
    boxPadding = 3,    -- Padding around the player model
    useCorners = true,  -- Use corners instead of full box
    cornerLength = 10   -- Length of corner lines when using corners
}

-- Pre-cache functions and variables
local RealTime = globals.RealTime
local TraceLine = engine.TraceLine
local GetByUserID = entities.GetByUserID
local GetLocalPlayer = entities.GetLocalPlayer
local targetData = {}
local lastLifeState = 2  -- Start with LIFE_DEAD (2)
local chamsMaterial = nil
local myfont = draw.CreateFont("Verdana", 16, 800)
local nextCleanupTime = 0
local nextVisCheck = {}
local visibilityCache = {}

-- Create materials
local function InitializeMaterials()
    if not chamsMaterial then
        chamsMaterial = materials.Create("targeted_chams", [[
            "VertexLitGeneric"
            {
                "$basetexture" "vgui/white_additive"
                "$bumpmap" "vgui/white_additive"
                "$color2" "[100 0.5 0.5]"
                "$selfillum" "1"
                "$ignorez" "1"
                "$selfIllumFresnel" "1"
                "$selfIllumFresnelMinMaxExp" "[0.1 0.2 0.3]"
                "$selfillumtint" "[0 0.3 0.6]"
            }
        ]])
    end
end

-- Initialize materials on script load
InitializeMaterials()

-- Improved visibility check with caching
local function IsPlayerVisible(player)
    if not CONFIG.visibleOnly then return true end
    
    local localPlayer = GetLocalPlayer()
    if not player or not localPlayer then return false end
    
    local id = player:GetIndex()
    local curTick = globals.TickCount()
    
    -- Use cached result if available
    if nextVisCheck[id] and curTick < nextVisCheck[id] then
        return visibilityCache[id] or false
    end
    
    -- Get eye position for more accurate tracing
    local localEyePos = localPlayer:GetAbsOrigin() + localPlayer:GetPropVector("localdata", "m_vecViewOffset[0]")
    local targetPos = player:GetAbsOrigin() + Vector3(0, 0, 40) -- Aim for chest area
    
    -- Use appropriate mask based on team
    local isTeammate = player:GetTeamNumber() == localPlayer:GetTeamNumber()
    local mask = isTeammate and MASK_SHOT_HULL or MASK_SHOT
    
    local trace = TraceLine(localEyePos, targetPos, mask)
    
    -- Cache result
    visibilityCache[id] = trace.entity == player or trace.fraction > 0.99
    nextVisCheck[id] = curTick + 2 -- Cache for 2 ticks
    
    return visibilityCache[id]
end

-- Draw corner lines
local function DrawCorners(x1, y1, x2, y2)
    local length = CONFIG.cornerLength
    
    -- Top left corner
    draw.Line(x1, y1, x1 + length, y1)  -- Horizontal
    draw.Line(x1, y1, x1, y1 + length)  -- Vertical
    
    -- Top right corner
    draw.Line(x2, y1, x2 - length, y1)  -- Horizontal
    draw.Line(x2, y1, x2, y1 + length)  -- Vertical
    
    -- Bottom left corner
    draw.Line(x1, y2, x1 + length, y2)  -- Horizontal
    draw.Line(x1, y2, x1, y2 - length)  -- Vertical
    
    -- Bottom right corner
    draw.Line(x2, y2, x2 - length, y2)  -- Horizontal
    draw.Line(x2, y2, x2, y2 - length)  -- Vertical
end

-- Clean expired attackers 
local function CleanExpiredAttackers(currentTime, targetInfo)
    local newAttackers = {}
    local count = 0
    local hasExpired = false
    
    for attackerIndex, timestamp in pairs(targetInfo.attackers) do
        if currentTime - timestamp <= CONFIG.trackerTimeWindow then
            newAttackers[attackerIndex] = timestamp
            count = count + 1
        else
            hasExpired = true
        end
    end
    
    targetInfo.attackerCount = count
    targetInfo.isMultiTargeted = (count >= CONFIG.minAttackers)
    
    return hasExpired and newAttackers or targetInfo.attackers
end

-- Clean up invalid targets
local function CleanInvalidTargets()
    local currentTime = RealTime()
    if currentTime < nextCleanupTime then return end
    
    nextCleanupTime = currentTime + CONFIG.cleanupInterval
    
    local localPlayer = GetLocalPlayer()
    if not localPlayer then return end
    
    for entIndex, targetInfo in pairs(targetData) do
        local entity = entities.GetByIndex(entIndex)
        
        -- Clean if entity is invalid, dead, dormant, or on our team
        if not entity or 
           not entity:IsValid() or 
           not entity:IsAlive() or
           entity:IsDormant() or
           entity:GetTeamNumber() == localPlayer:GetTeamNumber() then
            targetData[entIndex] = nil
            nextVisCheck[entIndex] = nil
            visibilityCache[entIndex] = nil
            goto continue
        end
        
        -- Update target status and clean if no longer multi-targeted
        targetInfo.attackers = CleanExpiredAttackers(currentTime, targetInfo)
        if targetInfo.attackerCount < CONFIG.minAttackers then
            targetData[entIndex] = nil
            nextVisCheck[entIndex] = nil
            visibilityCache[entIndex] = nil
        end
        
        ::continue::
    end
end

-- Reset all data (used on respawn)
local function ResetTargetData()
    targetData = {}
    nextVisCheck = {}
    visibilityCache = {}
    nextCleanupTime = 0
end

-- Check if player should be visualized
local function ShouldVisualizePlayer(player)
    if not player or 
       not player:IsValid() or 
       not player:IsAlive() or
       player:IsDormant() then 
        return false 
    end
    
    local localPlayer = GetLocalPlayer()
    if not localPlayer or player:GetTeamNumber() == localPlayer:GetTeamNumber() then 
        return false 
    end
    
    local targetInfo = targetData[player:GetIndex()]
    return targetInfo and targetInfo.isMultiTargeted and IsPlayerVisible(player)
end

-- Damage event handler
local function OnPlayerHurt(event)
    if event:GetName() ~= 'player_hurt' then return end
    
    local victim = GetByUserID(event:GetInt("userid"))
    local attacker = GetByUserID(event:GetInt("attacker"))
    
    if not victim or not attacker or victim == attacker then return end
    
    -- Skip dormant entities
    if victim:IsDormant() or attacker:IsDormant() then return end
    
    local victimIndex = victim:GetIndex()
    if not targetData[victimIndex] then
        targetData[victimIndex] = {
            attackers = {},
            isMultiTargeted = false,
            attackerCount = 0
        }
    end
    
    local currentTime = RealTime()
    local targetInfo = targetData[victimIndex]
    
    -- Update attackers list
    targetInfo.attackers[attacker:GetIndex()] = currentTime
    CleanExpiredAttackers(currentTime, targetInfo)
end

-- Main draw function
local function OnDrawModel(ctx)
    local entity = ctx:GetEntity()
    if not entity or not entity:IsPlayer() then return end
    
    -- Skip dormant entities
    if entity:IsDormant() then return end
    
    if ShouldVisualizePlayer(entity) then
        -- Apply chams if enabled
        if CONFIG.enableChams then
            ctx:ForcedMaterialOverride(chamsMaterial)
        end
    end
end

-- Draw border box around targeted players
local function DrawBox()
    if not CONFIG.enableBox then return end
    
    -- Set up drawing
    draw.SetFont(myfont)
    draw.Color(CONFIG.boxColor.r, CONFIG.boxColor.g, CONFIG.boxColor.b, CONFIG.boxColor.a)
    
    -- Find players being targeted by multiple teammates
    for index, targetInfo in pairs(targetData) do
        local player = entities.GetByIndex(index)
        if player and ShouldVisualizePlayer(player) then
            -- Get player bounds in screen space
            local origin = player:GetAbsOrigin()
            local mins = player:GetMins()
            local maxs = player:GetMaxs()
            
            -- Calculate corners in world space
            local corners = {
                Vector3(origin.x + mins.x - CONFIG.boxPadding, origin.y + mins.y - CONFIG.boxPadding, origin.z + mins.z),
                Vector3(origin.x + maxs.x + CONFIG.boxPadding, origin.y + mins.y - CONFIG.boxPadding, origin.z + mins.z),
                Vector3(origin.x + maxs.x + CONFIG.boxPadding, origin.y + maxs.y + CONFIG.boxPadding, origin.z + mins.z),
                Vector3(origin.x + mins.x - CONFIG.boxPadding, origin.y + maxs.y + CONFIG.boxPadding, origin.z + mins.z),
                Vector3(origin.x + mins.x - CONFIG.boxPadding, origin.y + mins.y - CONFIG.boxPadding, origin.z + maxs.z + CONFIG.boxPadding),
                Vector3(origin.x + maxs.x + CONFIG.boxPadding, origin.y + mins.y - CONFIG.boxPadding, origin.z + maxs.z + CONFIG.boxPadding),
                Vector3(origin.x + maxs.x + CONFIG.boxPadding, origin.y + maxs.y + CONFIG.boxPadding, origin.z + maxs.z + CONFIG.boxPadding),
                Vector3(origin.x + mins.x - CONFIG.boxPadding, origin.y + maxs.y + CONFIG.boxPadding, origin.z + maxs.z + CONFIG.boxPadding)
            }
            
            -- Convert to screen coordinates
            local screenCorners = {}
            for _, corner in ipairs(corners) do
                local screen = client.WorldToScreen(corner)
                if screen then
                    table.insert(screenCorners, screen)
                end
            end
            
            -- Draw the border if we have screen coordinates
            if #screenCorners > 0 then
                -- Find bounds
                local minX, minY = math.huge, math.huge
                local maxX, maxY = -math.huge, -math.huge
                
                for _, screen in ipairs(screenCorners) do
                    minX = math.min(minX, screen[1])
                    minY = math.min(minY, screen[2])
                    maxX = math.max(maxX, screen[1])
                    maxY = math.max(maxY, screen[2])
                end
                
                -- Draw either corners or full box based on config
                if CONFIG.useCorners then
                    -- Draw corners with thickness
                    for i = 0, CONFIG.boxThickness do
                        DrawCorners(minX - i, minY - i, maxX + i, maxY + i)
                    end
                else
                    -- Draw full box with thickness
                    for i = 0, CONFIG.boxThickness do
                        draw.OutlinedRect(minX - i, minY - i, maxX + i, maxY + i)
                    end
                end
            end
        end
    end
end

-- Monitor respawns to reset data
local function OnDraw()
    local localPlayer = GetLocalPlayer()
    if not localPlayer then return end
    
    local currentLifeState = localPlayer:GetPropInt("m_lifeState")
    if lastLifeState == 2 and currentLifeState == 0 then
        ResetTargetData()
    end
    lastLifeState = currentLifeState
    
    -- Clean up invalid targets periodically
    CleanInvalidTargets()
    
    -- Draw box effects
    DrawBox()
end

-- Register callbacks
callbacks.Register("FireGameEvent", "MultiTargetTracker", OnPlayerHurt)
callbacks.Register("DrawModel", "MultiTargetChams", OnDrawModel)
callbacks.Register("Draw", "MultiTargetEffects", OnDraw)
callbacks.Register("Unload", "MultiTargetCleanup", function()
    ResetTargetData()
    chamsMaterial = nil
end)

File: Amalgam/src/Features/Visuals/FocusFire/FocusFire.cpp
#include "FocusFire.h"
#include "../../../SDK/SDK.h"

void CFocusFire::Draw()
{
    if (!Vars::Competitive::Features::FocusFire.Value)
        return;
    
    // Early exit if game UI is visible
    if (I::EngineVGui->IsGameUIVisible())
        return;
    
    auto pLocal = H::Entities.GetLocal();
    if (!pLocal)
        return;
    
    // Clean up invalid targets periodically
    CleanInvalidTargets();
    
    // Draw box effects
    DrawBox();
}

bool CFocusFire::IsPlayerVisible(CTFPlayer* pPlayer)
{
    if (!Vars::Competitive::FocusFire::VisibleOnly.Value)
        return true;
    
    auto pLocal = H::Entities.GetLocal();
    if (!pPlayer || !pLocal)
        return false;
    
    int id = pPlayer->entindex();
    float curTime = I::GlobalVars->curtime;
    
    // Use cached result if available
    if (m_NextVisCheck.find(id) != m_NextVisCheck.end() && curTime < m_NextVisCheck[id])
        return m_VisibilityCache[id];
    
    // Get eye position for more accurate tracing
    Vec3 localEyePos = pLocal->GetEyePosition();
    Vec3 targetPos = pPlayer->GetAbsOrigin() + Vec3(0, 0, 40); // Aim for chest area
    
    // Use appropriate mask based on team
    bool isTeammate = pPlayer->m_iTeamNum() == pLocal->m_iTeamNum();
    int mask = isTeammate ? MASK_SHOT_HULL : MASK_SHOT;
    
    CGameTrace trace;
    CTraceFilterHitscan filter;
    filter.pSkip = pLocal;
    SDK::Trace(localEyePos, targetPos, mask, &filter, &trace);
    
    // Cache result
    m_VisibilityCache[id] = trace.m_pEnt == pPlayer || trace.fraction > 0.99f;
    m_NextVisCheck[id] = curTime + 0.1f; // Cache for 0.1 seconds
    
    return m_VisibilityCache[id];
}

void CFocusFire::DrawCorners(int x1, int y1, int x2, int y2)
{
    int length = Vars::Competitive::FocusFire::CornerLength.Value;
    
    // Top left corner
    H::Draw.Line(x1, y1, x1 + length, y1, Vars::Competitive::FocusFire::Color.Value);
    H::Draw.Line(x1, y1, x1, y1 + length, Vars::Competitive::FocusFire::Color.Value);
    
    // Top right corner
    H::Draw.Line(x2, y1, x2 - length, y1, Vars::Competitive::FocusFire::Color.Value);
    H::Draw.Line(x2, y1, x2, y1 + length, Vars::Competitive::FocusFire::Color.Value);
    
    // Bottom left corner
    H::Draw.Line(x1, y2, x1 + length, y2, Vars::Competitive::FocusFire::Color.Value);
    H::Draw.Line(x1, y2, x1, y2 - length, Vars::Competitive::FocusFire::Color.Value);
    
    // Bottom right corner
    H::Draw.Line(x2, y2, x2 - length, y2, Vars::Competitive::FocusFire::Color.Value);
    H::Draw.Line(x2, y2, x2, y2 - length, Vars::Competitive::FocusFire::Color.Value);
}

void CFocusFire::CleanExpiredAttackers(float currentTime, TargetInfo& targetInfo)
{
    std::unordered_map<int, float> newAttackers;
    int count = 0;
    
    for (auto& [attackerIndex, timestamp] : targetInfo.Attackers)
    {
        if (currentTime - timestamp <= Vars::Competitive::FocusFire::TrackerTimeWindow.Value)
        {
            newAttackers[attackerIndex] = timestamp;
            count++;
        }
    }
    
    targetInfo.Attackers = newAttackers;
    targetInfo.AttackerCount = count;
    targetInfo.IsMultiTargeted = (count >= Vars::Competitive::FocusFire::MinAttackers.Value);
}

void CFocusFire::CleanInvalidTargets()
{
    float currentTime = I::GlobalVars->curtime;
    if (currentTime < m_NextCleanupTime)
        return;
    
    m_NextCleanupTime = currentTime + CLEANUP_INTERVAL;
    
    auto pLocal = H::Entities.GetLocal();
    if (!pLocal)
        return;
    
    auto it = m_TargetData.begin();
    while (it != m_TargetData.end())
    {
        int entIndex = it->first;
        TargetInfo& targetInfo = it->second;
        
        auto pEntity = I::ClientEntityList->GetClientEntity(entIndex);
        auto pPlayer = pEntity ? pEntity->As<CTFPlayer>() : nullptr;
        
        // Clean if entity is invalid, dead, dormant, or on our team
        if (!pPlayer || 
            !pPlayer->IsAlive() || 
            pPlayer->IsDormant() || 
            pPlayer->m_iTeamNum() == pLocal->m_iTeamNum())
        {
            m_NextVisCheck.erase(entIndex);
            m_VisibilityCache.erase(entIndex);
            it = m_TargetData.erase(it);
            continue;
        }
        
        // Update target status and clean if no longer multi-targeted
        CleanExpiredAttackers(currentTime, targetInfo);
        if (targetInfo.AttackerCount < Vars::Competitive::FocusFire::MinAttackers.Value)
        {
            m_NextVisCheck.erase(entIndex);
            m_VisibilityCache.erase(entIndex);
            it = m_TargetData.erase(it);
            continue;
        }
        
        ++it;
    }
}

bool CFocusFire::ShouldVisualizePlayer(CTFPlayer* pPlayer)
{
    if (!pPlayer || 
        !pPlayer->IsAlive() || 
        pPlayer->IsDormant())
        return false;
    
    auto pLocal = H::Entities.GetLocal();
    if (!pLocal || pPlayer->m_iTeamNum() == pLocal->m_iTeamNum())
        return false;
    
    auto it = m_TargetData.find(pPlayer->entindex());
    return it != m_TargetData.end() && it->second.IsMultiTargeted && IsPlayerVisible(pPlayer);
}

void CFocusFire::DrawBox()
{
    if (!Vars::Competitive::FocusFire::EnableBox.Value)
        return;
    
    
    // Find players being targeted by multiple teammates
    for (auto& [index, targetInfo] : m_TargetData)
    {
        auto pEntity = I::ClientEntityList->GetClientEntity(index);
        auto pPlayer = pEntity ? pEntity->As<CTFPlayer>() : nullptr;
        
        if (pPlayer && ShouldVisualizePlayer(pPlayer))
        {
            // Chams are now handled in UpdateChamsEntities()
            
            // Skip box drawing if disabled
            if (!Vars::Competitive::FocusFire::EnableBox.Value)
                continue;
            // Get player bounds in world space
            Vec3 origin = pPlayer->GetAbsOrigin();
            Vec3 mins = pPlayer->m_vecMins();
            Vec3 maxs = pPlayer->m_vecMaxs();
            
            // Calculate corners in world space
            Vec3 corners[8] = {
                Vec3(origin.x + mins.x - Vars::Competitive::FocusFire::BoxPadding.Value, origin.y + mins.y - Vars::Competitive::FocusFire::BoxPadding.Value, origin.z + mins.z),
                Vec3(origin.x + maxs.x + Vars::Competitive::FocusFire::BoxPadding.Value, origin.y + mins.y - Vars::Competitive::FocusFire::BoxPadding.Value, origin.z + mins.z),
                Vec3(origin.x + maxs.x + Vars::Competitive::FocusFire::BoxPadding.Value, origin.y + maxs.y + Vars::Competitive::FocusFire::BoxPadding.Value, origin.z + mins.z),
                Vec3(origin.x + mins.x - Vars::Competitive::FocusFire::BoxPadding.Value, origin.y + maxs.y + Vars::Competitive::FocusFire::BoxPadding.Value, origin.z + mins.z),
                Vec3(origin.x + mins.x - Vars::Competitive::FocusFire::BoxPadding.Value, origin.y + mins.y - Vars::Competitive::FocusFire::BoxPadding.Value, origin.z + maxs.z + Vars::Competitive::FocusFire::BoxPadding.Value),
                Vec3(origin.x + maxs.x + Vars::Competitive::FocusFire::BoxPadding.Value, origin.y + mins.y - Vars::Competitive::FocusFire::BoxPadding.Value, origin.z + maxs.z + Vars::Competitive::FocusFire::BoxPadding.Value),
                Vec3(origin.x + maxs.x + Vars::Competitive::FocusFire::BoxPadding.Value, origin.y + maxs.y + Vars::Competitive::FocusFire::BoxPadding.Value, origin.z + maxs.z + Vars::Competitive::FocusFire::BoxPadding.Value),
                Vec3(origin.x + mins.x - Vars::Competitive::FocusFire::BoxPadding.Value, origin.y + maxs.y + Vars::Competitive::FocusFire::BoxPadding.Value, origin.z + maxs.z + Vars::Competitive::FocusFire::BoxPadding.Value)
            };
            
            // Convert to screen coordinates
            std::vector<Vec3> screenCorners;
            for (int i = 0; i < 8; i++)
            {
                Vec3 screen;
                if (SDK::W2S(corners[i], screen))
                    screenCorners.push_back(screen);
            }
            
            // Draw the border if we have screen coordinates
            if (!screenCorners.empty())
            {
                // Find bounds
                float minX = screenCorners[0].x, minY = screenCorners[0].y;
                float maxX = screenCorners[0].x, maxY = screenCorners[0].y;
                
                for (const auto& screen : screenCorners)
                {
                    minX = std::min(minX, screen.x);
                    minY = std::min(minY, screen.y);
                    maxX = std::max(maxX, screen.x);
                    maxY = std::max(maxY, screen.y);
                }
                
                // Draw either corners or full box based on config
                if (Vars::Competitive::FocusFire::UseCorners.Value)
                {
                    // Draw corners with thickness
                    for (int i = 0; i < Vars::Competitive::FocusFire::BoxThickness.Value; i++)
                    {
                        DrawCorners(static_cast<int>(minX - i), static_cast<int>(minY - i), 
                                   static_cast<int>(maxX + i), static_cast<int>(maxY + i));
                    }
                }
                else
                {
                    // Draw full box with thickness
                    for (int i = 0; i < Vars::Competitive::FocusFire::BoxThickness.Value; i++)
                    {
                        H::Draw.LineRect(static_cast<int>(minX - i), static_cast<int>(minY - i), 
                                        static_cast<int>(maxX + i - minX + 2*i), static_cast<int>(maxY + i - minY + 2*i), 
                                        Vars::Competitive::FocusFire::Color.Value);
                    }
                }
            }
        }
    }
}

void CFocusFire::OnPlayerHurt(int victimIndex, int attackerIndex)
{
    
    auto pVictim = I::ClientEntityList->GetClientEntity(victimIndex);
    auto pAttacker = I::ClientEntityList->GetClientEntity(attackerIndex);
    
    if (!pVictim || !pAttacker)
        return;
    
    auto pVictimPlayer = pVictim->As<CTFPlayer>();
    auto pAttackerPlayer = pAttacker->As<CTFPlayer>();
    
    if (!pVictimPlayer || !pAttackerPlayer)
        return;
    
    auto pLocal = H::Entities.GetLocal();
    if (!pLocal)
        return;
    
    // Skip if victim == attacker (self-damage)
    if (victimIndex == attackerIndex)
        return;
    
    // Skip if attacker is on same team as victim (friendly fire shouldn't count)
    if (pVictimPlayer->m_iTeamNum() == pAttackerPlayer->m_iTeamNum())
        return;
    
    // Skip if attacker is not on our team (we only track our team's attacks)
    if (pAttackerPlayer->m_iTeamNum() != pLocal->m_iTeamNum())
        return;
    
    // Skip dormant entities
    if (pVictimPlayer->IsDormant() || pAttackerPlayer->IsDormant())
        return;
    
    if (m_TargetData.find(victimIndex) == m_TargetData.end())
    {
        m_TargetData[victimIndex] = TargetInfo();
    }
    
    float currentTime = I::GlobalVars->curtime;
    TargetInfo& targetInfo = m_TargetData[victimIndex];
    
    // Update attackers list
    targetInfo.Attackers[attackerIndex] = currentTime;
    CleanExpiredAttackers(currentTime, targetInfo);
    
}

void CFocusFire::Reset()
{
    m_TargetData.clear();
    m_NextVisCheck.clear();
    m_VisibilityCache.clear();
    m_NextCleanupTime = 0.0f;
}

void CFocusFire::UpdateChamsEntities()
{
    // Clear previous chams entries
    m_mEntities.clear();
    
    if (!Vars::Competitive::Features::FocusFire.Value || !Vars::Competitive::FocusFire::EnableChams.Value)
        return;
    
    // Clean up invalid targets first
    CleanInvalidTargets();
    
    // Find players being targeted by multiple teammates
    for (auto& [index, targetInfo] : m_TargetData)
    {
        auto pEntity = I::ClientEntityList->GetClientEntity(index);
        auto pPlayer = pEntity ? pEntity->As<CTFPlayer>() : nullptr;
        
        if (pPlayer && ShouldVisualizePlayer(pPlayer))
        {
            // Add to chams tracking
            m_mEntities[pPlayer->entindex()] = true;
        }
    }
}